### PR TITLE
Support Enterprise Distribution

### PR DESCRIPTION
**Pull request by Automatt**
When distributing apps using an Enterprise certificate, there is no entry in the App Store from which to obtain the current version.  This adds the ability to load the current version from the distribution manifest file.

With enterprise distribution, the archived app is installed via an OTA installer generated by Xcode.  It consists of a manifest plist file that is stored on an accessible server, and a specific link that points to that file.  Since the manifest file contains the bundle version, it makes sense to load the current version of the app from there.

This adds a setting and enumeration SirenVersionCheckMethod that defaults to .AppStore but also supports .Manifest(manifestUrl: String).  Using this setting, a developer may use enterprise distribution during alpha testing and then switch to using the App Store as the source of truth for version numbers after the app ships.

### CODE REVIEW

File: Siren/Siren.swift [hunk 1]
```diff
@@ -50,6 +50,14 @@ public enum SirenVersionCheckType: Int {
 }
 
 /**
+    Determines the method used to perform the version check
+*/
+public enum SirenVersionCheckMethod {
+    case AppStore           // Use the version available in the app store (default)
+    case Manifest(String)   // Use the version described in a manifest plist (for enterprise distribution)
```
Comment (getaaron) [inline Siren/Siren.swift]: I like this use of enums here :+1:

File: Siren/Siren.swift [hunk 2]
```diff
@@ -265,6 +276,76 @@ public class Siren: NSObject {
...
     
     private func performVersionCheck() {
+        switch checkMethod {
+            
+        case .AppStore:
+            performVersionCheckAppStore()
+            
+        case .Manifest(let manifestUri):
+            performVersionCheckManifest(manifestUri)
+        }
```
Comment (getaaron) [inline Siren/Siren.swift]: - Not sure I like calling it "Uri". I think this should be uppercase, and URL, to match other variables like `storeURLString`
- Instead of having two clusters of similar methods, this might be good opportunity to define a protocol, and have two different structs which can both implement `performVersionCheck()`

File: Siren/Siren.swift [hunk 3]
```diff
@@ -265,6 +276,76 @@ public class Siren: NSObject {
...
+    
+    private func performVersionCheckManifest(manifestUri: String) {
+        
+        let manifestURL = manifestURLFromString(manifestUri)
+        let request = NSMutableURLRequest(URL: manifestURL)
+        request.HTTPMethod = "GET"
+        
+        // Perform Request
+        let session = NSURLSession.sharedSession()
+        let task = session.dataTaskWithRequest(request, completionHandler: { (data, response, error) -> Void in
```
Comment (getaaron) [inline Siren/Siren.swift]: I think this would look cleaner with trailing closure syntax:

    let task = session.dataTaskWithRequest(request) {

File: Siren/Siren.swift [hunk 4]
```diff
@@ -265,6 +276,76 @@ public class Siren: NSObject {
...
+            
+            if let error = error {
```
Comment (getaaron) [inline Siren/Siren.swift]: I think it would make more sense to `guard let response = response` and  only unwrap the error in the `else` block. This would keep the "happy path" code at a higher level and the error handling stuff indented.

File: Siren/Siren.swift [hunk 5]
```diff
@@ -265,6 +276,76 @@ public class Siren: NSObject {
...
+                        if self.debugEnabled {
+                            print("[Siren] plist results: \(appData)")
+                        }
+                        
+                        // Process Results (e.g., extract current version on the AppStore)
+                        self.processManifestVersionCheckResults(appData)
+                        
+                    })
+                    
+                } catch let error as NSError {
```
Comment (getaaron) [inline Siren/Siren.swift]: This can just be `catch` (you don't need the `let error as NSError`). Then you can also drop `.localizedDescription` in the print statement.

File: Siren/Siren.swift [hunk 6]
```diff
@@ -322,6 +403,51 @@ public class Siren: NSObject {
         task.resume()
     }
     
+    private func processManifestVersionCheckResults(lookupResults: [String: AnyObject]) {
+        
+        // Store version comparison date
+        storeVersionCheckDate()
+        
+        guard let items = lookupResults["items"] as? Array<AnyObject> else {
```
Comment (getaaron) [inline Siren/Siren.swift]: :+1: Great use of `guard`. I think I'd use `[AnyObject]` instead of the longer `Array<AnyObject>`

File: Siren/Siren.swift [hunk 7]
```diff
@@ -322,6 +403,51 @@ public class Siren: NSObject {
...
+            if debugEnabled {
+                print("[Siren] Error retrieving manifest as there was no data returned")
+            }
+            return
+        }
+        
+        guard let bundle = items[0] as? [String: AnyObject] else {
```
Comment (getaaron) [inline Siren/Siren.swift]: Since all these error messages are identical, should we combine them into one `guard` statement? `guard let items = …, let bundle = …, let metadata = …, let version = … else { }`

It would certainly make the code shorter, although it might be harder to read.

File: Siren/Siren.swift [hunk 8]
```diff
@@ -322,6 +403,51 @@ public class Siren: NSObject {
...
+            if debugEnabled {
+                print("[Siren] Error retrieving manifest version as there was no data returned")
+            }
+            return
+        }
+        
+        currentAppStoreVersion = version
+        if isAppStoreVersionNewer() {
+            showAlertIfCurrentAppStoreVersionNotSkipped()
+        } else {
```
Comment (getaaron) [inline Siren/Siren.swift]: We could shorten this to just `else if debugEnabled`

**Comment by getaaron** [comment]
Wonderful feature and PR, @Automatt. I left a few stylistic suggestions; let me know what you think. CC: @ArtSabintsev

File: Siren/Siren.swift (hunk 4 above)
Comment (Automatt) [inline Siren/Siren.swift]: I agree, great suggestion.

File: Siren/Siren.swift (hunk 7 above)
Comment (Automatt) [inline Siren/Siren.swift]: Chaining these together does look a lot cleaner, thanks!

File: Siren/Siren.swift (hunk 2 above)
Comment (Automatt) [inline Siren/Siren.swift]: I agree having two clusters of similar methods is not ideal.  I'll do a pass using a protocol -- I haven't used protocols nearly enough so I'd love to get your feedback on that.  Thanks!

**Comment by ArtSabintsev** [comment]
Will look into this after work today. Thanks for the PR!

**Comment by ArtSabintsev** [comment]
Sorry, I've been a bit busy. This will have to wait a couple more days.

**Comment by ArtSabintsev** [comment]
Hey - sorry, still a bit backlogged. Traveling a bit this week + in the midst of the push.

I will tell you up front that I do have some reservations about adding this code in the main branch, but I'll save them until after I review everything.

**Comment by Automatt** [comment]
Thanks, no worries either way.

—
Mathew Spolin
matt@automatt.com
415-509-1199

**Comment by ArtSabintsev** [comment]
hey @Automatt, and by extension, @getaaron,

I think it'd be best if a new open source library be written for enterprise updates. I'm more than happy to work on this with you. On top of that, this would give me a reason to split out the language localizations from harpy.bundle and siren.bundle, and put them into their own pod that all three libraries would depend on.

What do you think?

**Comment by Automatt** [comment]
I like that idea.  Most apps are probably either distributed using Enterprise OR App Store.  Mine has a weird requirement where it is expected to be both, so it's important for me to be able to switch without significant changes to the project.

**Comment by ArtSabintsev** [comment]
I've just spent a ton of time this past week with Enterprise Apps. I can definitely see this feature, however, I stand firm with my previous statement that it should be managed in a separate repo.

I'm closing this issue for now, but if you want, I can work on putting together a separate library.
